PEPPER-587: RGP DSM playwright tests_assert creation of new participant by checking DSM

Rinchan20 opened the pull request:
Extends the current RGP playwright test to check that the newly created participant can also be seen in DSM - might need a slight tweak - the most recent changes were done using DSM Test and there's a slight text difference when selecting the RGP study there (DSM Test has `Pepper RGP`, DSM Dev has `RGP`). But other than that should be running fine. 

Finishes up the last part of of this ticket: https://broadworkbench.atlassian.net/browse/PEPPER-587

Rinchan20 committed:
PR for checking the new RGP pt is located in DSM

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

Rinchan20 committed:
Update adult-enrollment.spec.ts

Changed from 'Pepper RGP' -> 'RGP' so that it'll be able to run no matter if used in dev, test, or staging

Rinchan20 committed:
ESLint errors fixed

Rinchan20 committed:
Update faker-utils.ts

missed an eslint error

Rinchan20 commented:
tweak from `Pepper RGP` -> `RGP` done as well

GiCharkviani commented on playwright-e2e/pages/dsm/home-page.ts:
```diff
@@ -1,13 +1,63 @@
+  }
+
+  async selectStudy(study: string, page: Page) {
+    const studySelector = await new Select(page, { label: 'Select study' });
```
Instead, you could use this way: `this.navigation.selectStudy()`

GiCharkviani replied:
Oh, anyway, this method is not used anywhere. Maybe remove it?

GiCharkviani commented on playwright-e2e/pages/dsm/home-page.ts:
```diff
@@ -1,13 +1,63 @@
+   * @param opts
+   */
+  async logIn(opts: { email?: string; password?: string; waitForNavigation?: boolean } = {}) {
+    await auth.login(this.page);
```
I think when you are on this page, you are already logged in, therefore I don't think this method is useful here. What do you think about removing it? 
Please correct me if I'm wrong.

Usually, we just call the `login() `method from `beforeEach`. 
It would be good if we considered later adding **Login Page Object Model**, which would handle all login-related things

GiCharkviani commented on playwright-e2e/pages/dsm/home-page.ts:
```diff
@@ -1,13 +1,63 @@
+    await studySelector.selectOption(study);
+  }
+
+  async selectStudyMenuOption(menuOption: string) {
```
This is not used as well

GiCharkviani commented on playwright-e2e/pages/dsm/home-page.ts:
```diff
@@ -1,13 +1,63 @@
+    }
+  }
+
+  async selectCustomizeViewOption(columnGroup: string, columnName: string) {
```
This is not used as well. 
If you need to use customize view functionality anywhere, you can find it in CustomizeView class, which is located in the Filters class, that is available (according to DSM structure) in the ParticipantListPage class (POM).

GiCharkviani commented on playwright-e2e/pages/dsm/participantList-page.ts:
```diff
@@ -20,6 +24,18 @@ export default class ParticipantListPage {
   }
 
+  public async filterListByParticipantGUID(participantGUID: string): Promise<void> {
+    await this.page.locator('text=Customize View >> button').click();
```
Instead, you could use customize view functionality, which is available from the `_filters` field. Like so:
```
const customizeViewPanel = this.filters.customizeViewPanel;
await customizeViewPanel.open();
await customizeViewPanel.selectColumns('Participant Columns', ['Participant ID']);

const searchPanel = this.filters.searchPanel;
await searchPanel.open();
await searchPanel.text('Participant ID', {textValue:participantGUID});
await searchPanel.search();
```

And later to make assertions.

What do you think?

Rinchan20 replied:
sounds good - will make the changes

GiCharkviani commented on playwright-e2e/tests/rgp/adult-enrollment.spec.ts:
```diff
@@ -9,8 +9,17 @@ import TellUsAboutYourFamilyPage from 'pages/rgp/enrollment/tell-us-about-your-f
-const { RGP_USER_EMAIL, RGP_USER_PASSWORD } = process.env;
+import { request } from 'http';
+import { updateTypeAssertion } from 'typescript';
+import { setPatientParticipantGuid } from 'utils/faker-utils';
```
Unused imports:

```
import { request } from 'http';
import { updateTypeAssertion } from 'typescript';
```

GiCharkviani commented on playwright-e2e/pages/dsm/home-page.ts:
```diff
@@ -1,13 +1,63 @@
+import { Navigation } from 'lib/component/dsm/navigation/navigation';
+import { StudyNav } from 'lib/component/dsm/navigation/enums/studyNav.enum';
+import ParticipantListPage from 'pages/dsm/participantList-page';
+import ParticipantPage from 'pages/dsm/participant-page';
```
Unused import:
`import ParticipantPage from 'pages/dsm/participant-page';
`

aweng98 commented on playwright-e2e/pages/dsm/home-page.ts:
```diff
@@ -1,13 +1,63 @@
+      await this.navigation.selectFromStudy<ParticipantListPage>(StudyNav.PARTICIPANT_LIST);
+      await this.participantList.waitForReady();
+      await this.participantList.assertPageTitle();
+      await this.participantList.waitForReady();
```
Line #47 is not needed. It's dupl of line #45. `assertPageTitle()` does not click.

aweng98 commented on playwright-e2e/pages/dsm/home-page.ts:
```diff
@@ -1,13 +1,63 @@
+    await this.page.locator(`text='${columnName}' >> button`).click();
+  }
+
+  async waitForReady(): Promise<void> {
+    await this.page.waitForFunction(() => document.title === 'DDP Study Management');
+  }
```
FYI, `document.title === 'DDP Study Managment'` tells me current page is the expected page. But it doesn't checks page is fully loaded and ready for user interaction. Just wanted you to know this, it's okay to leave it as is.

aweng98 commented on playwright-e2e/pages/dsm/page-base.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Data Study Manager (DSM) page base
+ */
+export abstract class DSMPageBase extends PageBase {
```
👍🏻👍🏻

aweng98 commented on playwright-e2e/pages/dsm/participantList-page.ts:
```diff
@@ -1,9 +1,13 @@
+  ParticipantID = 'Participant ID'
+}
+
 export default class ParticipantListPage {
```
`ParticipantListPage extends DSMPageBase`? because you created new abstract class `DSMPageBase`.

aweng98 commented on playwright-e2e/utils/faker-utils.ts:
```diff
@@ -17,3 +19,13 @@ export const generateEmailAlias = (email: string | undefined): string => {
   return `${name}+${Math.floor(Math.random() * 1000000000)}@${domain}`;
 };
+
+export const setPatientParticipantGuid = async (page: Page) => {
+  const [response] = await Promise.all([
+    page.waitForResponse((resp) => resp.url().includes('/participants') && resp.status() === 200)
+  ]);
+
+  const participantResponse = response.url();
+  const urlArray = participantResponse.split('/');
+  user.patient.participantGuid = urlArray[6];
+};
```
Waiting and parsing a request in a separate function is going to make tests flaky because request could have been finished before this function starts to run. Best way to do this is, perform action and wait for response together a in Promise.all.

```
await Promise.all([
  some_action(),
  waitForResponse()
]);
```

Rinchan20 replied:
will do - thinking a bit more about this one - in terms of which action to use - will make the changes

aweng98 commented on playwright-e2e/tests/rgp/adult-enrollment.spec.ts:
```diff
@@ -152,5 +164,22 @@ test.describe('Adult Self Enrollment', () => {
     expect(await tellUsAboutYourFamily.yourTitle().isDisabled()).toEqual(true);
     expect(await tellUsAboutYourFamily.yourFirstName().isDisabled()).toEqual(true);
+    //Go to DSM to verify the newly created account can be found there
+    const dsm = new dsmHome(page);
```
Recommendation: Make a test to do dsm checking. That way, if DSM checking fails, RGP test is not affected.
```
test('Go to DSM to verify the newly created account can be found @functional @enrollment @rgp')
```

Rinchan20 committed:
implementing improvement suggestions

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

aweng98 commented on playwright-e2e/tests/rgp/adult-enrollment.spec.ts:
```diff
@@ -9,10 +9,17 @@ import TellUsAboutYourFamilyPage from 'pages/rgp/enrollment/tell-us-about-your-f
+const { RGP_USER_EMAIL, RGP_USER_PASSWORD, DSM_USER_EMAIL, DSM_USER_PASSWORD } = process.env;
 
-test.describe('Adult Self Enrollment', () => {
+test.describe.serial('Adult Self Enrollment', () => {
```
Correct!

Rinchan20 replied:
Will take a look at the test and fix it post haste

Rinchan20 replied:
initial investigation - seems to run well in my localhost - but for some reason the pt above is in the dev database but doesn't seem to exist in elastic search - looking more into it and will report back

Rinchan20 replied:
kibana seems to be slow - so it might be more than a couple of mins

Rinchan20 replied:
correction - I used test env when checking local run - that's why it's working for me - new RGP pts since about March 10th don't seem to be appearing in DSM and that's why the test failed - looking into it

Rinchan20 replied:
*don't seem to be appearing in DSM dev I mean

aweng98 commented on playwright-e2e/package-lock.json:
```diff
@@ -31,6 +30,9 @@
         "@typescript-eslint/parser": "^5.54.0",
         "eslint": "^8.35.0",
         "eslint-plugin-import": "^2.27.5",
+        "eslint-plugin-prettier": "^4.2.1",
+        "prettier": "^2.8.4",
```
I have removed prettier and related plugin. Do you think we want to keep it?

Rinchan20 replied:
no - let's take it out (will do!) - it ended up getting put in while using vs code

aweng98 reviewed (commented):
Test failed on my localhost. Can you take a look?
<img width="1082" alt="Screenshot 2023-03-16 at 1 06 09 PM" src="https://user-images.githubusercontent.com/35533885/225697577-2329a886-3f17-4b26-9301-2fc536725e61.png">

Rinchan20 committed:
WIP

filling out more proband automation in prep for family history automation

Rinchan20 committed:
contact info automated

Rinchan20 committed:
study status automated

Rinchan20 committed:
medical records automated

Rinchan20 committed:
primary sample automated

Rinchan20 committed:
tissue section in progress

Rinchan20 committed:
tissue section automated

Rinchan20 committed:
checking family ids WIP

Rinchan20 committed:
added initial add family member automation

also implemented some feedback

Rinchan20 committed:
added separate test for non copied family member

Rinchan20 committed:
implemented feedback

Rinchan20 committed:
a few more assertions

Rinchan20 committed:
slight tweaks and validation for textareas

there was a bug some time ago where inputs to textareas would not be saved or would be missing upon revisiting participant page - adding assertions to check that input for previously affected fields are saved

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

Rinchan20 committed:
some tweaks with textarea automation wip

Rinchan20 committed:
tweaks and textcontent automation done

Rinchan20 commented:
remaining work: some feedback, comparison between copied family member and proband (DSM), checking messages display (DSS) and then should be end to end. Will also add blood rnaseq column check. Then most apart from kit upload for RGP should be automated with this PR.

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

Rinchan20 committed:
Update adult-enrollment.spec.ts

flakiness in checking contents of Participant Info notes after revisiting page seems to be mostly fixed

Rinchan20 commented:
PR currently adds the last 5 tests seen in the below photo - the test ("Verify that a family member can be added using copied proband info") is halfway done - copied family member is made, comparison to the proband just needs to be done - currently running on test due to dev housekeeping issues:
<img width="655" alt="Screen Shot 2023-05-24 at 12 12 58 PM" src="https://github.com/broadinstitute/ddp-angular/assets/5404082/f174bbe0-1561-4d1c-8ef0-659e439cf003">
(not sure if photo of tests with runtimes will display in github)

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

Rinchan20 committed:
moved things to match develop update

Rinchan20 committed:
fixed typo in family member enums

Rinchan20 committed:
added extra check for non-copied family member

Rinchan20 committed:
tweaks

move out most dsm related tests to its own script since test run time was approaching 180 seconds which is temporarily extended while doing testing - test time for dsm related stuff is now 70 seconds - will likely combine a test or two - other tests are skipped so that a run focuses on only adult enrollment and dsm - also when child enrollment fails screenshot comparison the guid info seems lost somehow - will get rid of skips once kit upload and dashboard part is done

Rinchan20 committed:
Create rgpFinalScan-page.ts

Rinchan20 committed:
rgp kit upload - work in progress

Rinchan20 committed:
added fs-extra in an attempt to get file location recognized

Rinchan20 committed:
WIP

Rinchan20 committed:
WIP

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

aweng98 commented on playwright-e2e/package.json:
```diff
@@ -16,15 +16,17 @@
   "devDependencies": {
     "@faker-js/faker": "^7.6.0",
-    "@playwright/test": "^1.31.1",
+    "@playwright/test": "^1.33.0",
```
Don't upgrade Playwright version in this PR because you would also need to update CI Playwright docker image version, update test screenshots on Dev, Test and docker container.

Rinchan20 committed:
WIP

Rinchan20 committed:
changes to get most recent playwright participant

Rinchan20 committed:
update tests to use new method for getting guids

Rinchan20 committed:
Merge branch 'develop' into rgp_dsm_playwright_tests

Rinchan20 committed:
kit upload part 99 percent done want to add more checks

currently rgp tests check the following: dss enrollment survey -> dsm to check the pt is there -> check the rgp dynamic fields -> family member creation -> kit upload. Added method to get most recent playwright pt. Added stuff from other pr since merge both option wasn't seen.

Rinchan20 committed:
some changes after some tests failed in docker

the @proband annotation is temporary while cleaning up tests

Rinchan20 committed:
participant guid changes

forgot to make family enrollment tests serial, the first test now searches for the most recently created automated pt and the rests of the tests after that should consistently use whatever pt the first test found. Could use refactoring later. My additions seem to pass in docker container now.

Rinchan20 committed:
some cleanup and adding results of screenshot comparison

Rinchan20 commented:
Plan is to break this PR into smaller PRs - a bit on hold until this ticket is done: https://broadworkbench.atlassian.net/browse/PEPPER-692 (since it will fail consistently in dev otherwise - currently passes when ran in tests which has less/minimal housekeeping issues)

Rinchan20 commented:
handled by other smaller PRs - closing this one

Rinchan20 commented:
handled by PRs such as:

- https://github.com/broadinstitute/ddp-angular/pull/2130
- https://github.com/broadinstitute/ddp-angular/pull/2160
- https://github.com/broadinstitute/ddp-angular/pull/2180
- https://github.com/broadinstitute/ddp-angular/pull/2181
- https://github.com/broadinstitute/ddp-angular/pull/2183
